pinot-core/src/main/java/org/apache/pinot/core/query/selection/SelectionOperatorUtils.java

```diff
@@ -481,55 +510,81 @@ public static ResultTable renderResultTableWithoutOrdering(List<Serializable[]>
   }
 
   /**
-   * Extract columns from the row based on the given column indices.
-   * <p>The extracted row is used to build the {@link SelectionResults}.
-   *
-   * @param row selection row to be extracted.
-   * @param columnIndices column indices.
-   * @return selection row.
+   * Converts a value into the given data type. (Broker side)
+   * <p>Actual value type can be different with data type passed in, but they must be type compatible.
    */
-  public static Serializable[] extractColumns(Serializable[] row, int[] columnIndices,
-      @Nullable DataSchema.ColumnDataType[] columnDataTypes) {
-    int numColumns = columnIndices.length;
-    Serializable[] extractedRow = new Serializable[numColumns];
-    if (columnDataTypes == null) {
-      for (int i = 0; i < numColumns; i++) {
-        extractedRow[i] = row[columnIndices[i]];
-      }
-    } else {
-      for (int i = 0; i < numColumns; i++) {
-        extractedRow[i] = getFormattedValue(row[columnIndices[i]], columnDataTypes[i]);
-      }
-    }
-    return extractedRow;
-  }
+  public static Serializable convertValueToType(Object value, DataSchema.ColumnDataType dataType) {
+    switch (dataType) {
+      // Single-value column
+      case INT:
+        return ((Number) value).intValue();
```

Comment:
From `IntermediateResultsBlock` class, you kind of directly casted types to an object instead of casting to `Number`.  Can we use the same approach here?

Reply:
We are trying to handle compatible types here (commented on line 514). This is required for cases where different segments have different schema (same column name, but different data types).
The long term solution should be let the query engine use the table schema instead of the schema inside each segments.